File: sacco-ui/src/services/investmentService.ts
import { apiRequest } from './apiClient'
import type {
  InvestmentResponse,
  CreateInvestmentRequest,
  InvestmentIncomeResponse,
  RecordIncomeRequest,
  InvestmentValuationResponse,
  RecordValuationRequest,
  DisposeInvestmentRequest,
  RollOverRequest,
  InvestmentSummary,
} from '../types/investments'

const BASE = '/api/v1/investments'

export type AuthenticatedRequest = <T>(path: string, options?: RequestInit) => Promise<T>

function callApi<T>(
  path: string,
  options: RequestInit | undefined,
  request?: AuthenticatedRequest,
): Promise<T> {
  if (request) return request<T>(path, options)
  return apiRequest<T>(path, options)
}

/* Portfolio */

export async function getInvestments(request?: AuthenticatedRequest): Promise<InvestmentResponse[]> {
  return callApi<InvestmentResponse[]>(BASE, undefined, request)
}

export async function getInvestment(id: string, request?: AuthenticatedRequest): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(`${BASE}/${id}`, undefined, request)
}

export async function getInvestmentSummary(request?: AuthenticatedRequest): Promise<InvestmentSummary> {
  return callApi<InvestmentSummary>(`${BASE}/summary`, undefined, request)
}

/* Lifecycle */

export async function createInvestment(
  payload: CreateInvestmentRequest,
  request?: AuthenticatedRequest,
): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(BASE, {
    method: 'POST',
    body: JSON.stringify(payload),
  }, request)
}

export async function approveInvestment(id: string, request?: AuthenticatedRequest): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(`${BASE}/${id}/approve`, { method: 'PATCH' }, request)
}

export async function rejectInvestment(
  id: string,
  reason?: string,
  request?: AuthenticatedRequest,
): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(`${BASE}/${id}/reject`, {
    method: 'PATCH',
    ...(reason ? { body: JSON.stringify({ reason }) } : {}),
  }, request)
}

export async function activateInvestment(id: string, request?: AuthenticatedRequest): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(`${BASE}/${id}/activate`, { method: 'PATCH' }, request)
}

export async function disposeInvestment(
  id: string,
  payload: DisposeInvestmentRequest,
  request?: AuthenticatedRequest,
): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(`${BASE}/${id}/dispose`, {
    method: 'POST',
    body: JSON.stringify(payload),
  }, request)
}

export async function rollOverInvestment(
  id: string,
  payload: RollOverRequest,
  request?: AuthenticatedRequest,
): Promise<InvestmentResponse> {
  return callApi<InvestmentResponse>(`${BASE}/${id}/rollover`, {
    method: 'POST',
    body: JSON.stringify(payload),
  }, request)
}

/* Income */

export async function getInvestmentIncome(
  investmentId: string,
  request?: AuthenticatedRequest,
): Promise<InvestmentIncomeResponse[]> {
  return callApi<InvestmentIncomeResponse[]>(`${BASE}/${investmentId}/income`, undefined, request)
}

export async function recordIncome(
  investmentId: string,
  payload: RecordIncomeRequest,
  request?: AuthenticatedRequest,
): Promise<InvestmentIncomeResponse> {
  return callApi<InvestmentIncomeResponse>(`${BASE}/${investmentId}/income`, {
    method: 'POST',
    body: JSON.stringify(payload),
  }, request)
}

/* Valuations */

export async function getValuations(
  investmentId: string,
  request?: AuthenticatedRequest,
): Promise<InvestmentValuationResponse[]> {
  return callApi<InvestmentValuationResponse[]>(`${BASE}/${investmentId}/valuations`, undefined, request)
}

export async function recordValuation(
  investmentId: string,
  payload: RecordValuationRequest,
  request?: AuthenticatedRequest,
): Promise<InvestmentValuationResponse> {
  return callApi<InvestmentValuationResponse>(`${BASE}/${investmentId}/valuations`, {
    method: 'POST',
    body: JSON.stringify(payload),
  }, request)
}

File: sacco-investment/src/main/java/com/innercircle/sacco/investment/service/InvestmentService.java
package com.innercircle.sacco.investment.service;

import com.innercircle.sacco.investment.dto.CreateInvestmentRequest;
import com.innercircle.sacco.investment.dto.DisposeInvestmentRequest;
import com.innercircle.sacco.investment.dto.InvestmentSummaryResponse;
import com.innercircle.sacco.investment.dto.RecordIncomeRequest;
import com.innercircle.sacco.investment.dto.RecordValuationRequest;
import com.innercircle.sacco.investment.dto.RollOverRequest;
import com.innercircle.sacco.investment.entity.Investment;
import com.innercircle.sacco.investment.entity.InvestmentIncome;
import com.innercircle.sacco.investment.entity.InvestmentValuation;

import java.util.List;
import java.util.UUID;

public interface InvestmentService {

    List<Investment> listInvestments();

    Investment getInvestment(UUID investmentId);

    InvestmentSummaryResponse getSummary();

    Investment createInvestment(CreateInvestmentRequest request, String actor);

    Investment approveInvestment(UUID investmentId, String actor);

    Investment rejectInvestment(UUID investmentId, String reason, String actor);

    Investment activateInvestment(UUID investmentId, String actor);

    Investment disposeInvestment(UUID investmentId, DisposeInvestmentRequest request, String actor);

    Investment rollOverInvestment(UUID investmentId, RollOverRequest request, String actor);

    List<InvestmentIncome> getIncomeHistory(UUID investmentId);

    InvestmentIncome recordIncome(UUID investmentId, RecordIncomeRequest request, String actor);

    List<InvestmentValuation> getValuationHistory(UUID investmentId);

    InvestmentValuation recordValuation(UUID investmentId, RecordValuationRequest request, String actor);
}
